### DISCUSSION TITLE
Can I get div's height/offsets before useTransition runs?

### DISCUSSION THREAD

**Discussion by mostrecent** [pmndrs/react-spring#864] [Support]
## 🤓 Question

I have a question re your list reordering example https://codesandbox.io/embed/1wqpz5mzqj

It needs pre-set heights in the data collection to run. How can I get it to work without setting the heights manually before? I thought about getting the div's offset just before useTransitions runs but the actuals div rely on it running first.

Somehow framer-motion manages this kind of animation without setting heights manually: https://codesandbox.io/s/framer-motion-reorder-animation-bviz6?fontsize=14&module=%2Fsrc%2FExample.tsx

**Comment by harshagarwalnyu** [comment]
Use `useMeasure` from `@react-spring/web` (or the standalone `react-use-measure`) to get real DOM dimensions before/during animation:

```jsx
import { useTransition, animated } from '@react-spring/web';
import useMeasure from 'react-use-measure';

function AnimatedList({ items }) {
  const [ref, { height }] = useMeasure();  // measures the list container

  const transitions = useTransition(items, {
    keys: (item) => item.id,
    from: { opacity: 0, height: 0 },
    enter: (item) => async (next) => {
      // height is measured automatically — no manual preset needed
      await next({ opacity: 1, height: item.measuredHeight });
    },
    leave: { opacity: 0, height: 0 },
  });

  return <div ref={ref}>{/* ... */}</div>;
}
```

**For the reorder list specifically** (like the CodeSandbox you linked), the modern pattern uses `useSpring` + `ResizeObserver` under the hood via `useMeasure`:

```jsx
import { useSprings, animated } from '@react-spring/web';
import useMeasure from 'react-use-measure';

function Item({ item }) {
  const [ref, { height }] = useMeasure();

  return (
    <animated.div style={{ height }}>
      <div ref={ref}>
        {item.content}
      </div>
    </animated.div>
  );
}
```

`useMeasure` uses `ResizeObserver` — it fires synchronously with layout so `height` is available before the first paint. You attach `ref` to the inner content div and animate the outer `animated.div` height from 0 → measured.

**How Framer Motion does it without manual heights:** Framer uses `layout` animations which hook into the browser's layout engine via FLIP (First, Last, Invert, Play). react-spring's equivalent is the `layout` prop added in newer versions, or you can implement FLIP manually with `getBoundingClientRect()` snapshots before/after state changes.

```bash
npm install react-use-measure
# or: @react-spring/web includes it
```
